--- input-libinput/man/libinput.man.orig	2025-07-25 14:34:09.447480278 +0300
+++ input-libinput/man/libinput.man	2025-07-25 14:42:24.000613113 +0300
@@ -178,6 +178,15 @@
 Enables middle button emulation.
 When enabled, pressing the left and right
 buttons simultaneously produces a middle mouse button click.
+A synonym
+.IP
+.BI "Option \*qEmulate3Buttons\*q \*q" bool \*q
+.IP
+is preserved for compatibility with
+.B evdev
+driver, with 
+.BI "\*qMiddleEmulation\*q
+having preference of both are set.
 .TP 7
 .BI "Option \*qNaturalScrolling\*q \*q" bool \*q
 Enables or disables natural scrolling behavior.
--- input-libinput/src/xf86libinput.c.orig	2025-07-25 14:32:15.471809962 +0300
+++ input-libinput/src/xf86libinput.c	2025-07-25 14:33:13.390010555 +0300
@@ -3489,13 +3489,17 @@
 xf86libinput_parse_middleemulation_option(InputInfoPtr pInfo,
 					  struct libinput_device *device)
 {
-	BOOL enabled;
+	int enabled;
 
 	if (!libinput_device_config_middle_emulation_is_available(device))
 		return FALSE;
 
 	enabled = xf86SetBoolOption(pInfo->options,
 				    "MiddleEmulation",
+				    -1); /* returns -1 if the option has not been set */
+	if (enabled == -1)
+		enabled = xf86SetBoolOption(pInfo->options,
+				    "Emulate3Buttons",
 				    libinput_device_config_middle_emulation_get_default_enabled(device));
 	if (libinput_device_config_middle_emulation_set_enabled(device, enabled) !=
 	    LIBINPUT_CONFIG_STATUS_SUCCESS) {
