--- xf86-video-ati-xlibre-xf86-video-ati-25.0.0/configure.ac.orig	2025-12-29 23:31:48.998861928 +0200
+++ xf86-video-ati-xlibre-xf86-video-ati-25.0.0/configure.ac	2025-12-29 23:31:58.021861613 +0200
@@ -24,7 +24,7 @@
 AC_PREREQ([2.60])
 AC_INIT([xlibre-xf86-video-ati],
         [25.0.0],
-        [https://github.com/X11Libre/xf86-video-ati/issues]
+        [https://github.com/X11Libre/xf86-video-ati/issues],
         [xlibre-xf86-video-ati])
 
 AC_CONFIG_SRCDIR([Makefile.am])
