--- xf86-input-wacom-xlibre-xf86-input-wacom-1.2.3.1/src/x11/xf86Wacom.c.orig	2025-09-17 15:45:29.297724701 +0300
+++ xf86-input-wacom-xlibre-xf86-input-wacom-1.2.3.1/src/x11/xf86Wacom.c	2025-09-17 15:47:36.419362752 +0300
@@ -71,7 +71,7 @@
 	va_list args;
 
 	va_start(args, format);
-	xf86VIDrvMsgVerb(priv->frontend, xtype, 0, format, args);
+	xf86IDrvMsgVerb(priv->frontend, xtype, 0, format, args);
 	va_end(args);
 }
 
