--- xserver-xlibre-xserver-25.1.0/hw/xfree86/loader/loader.c.orig	2025-12-25 23:53:55.112292868 +0200
+++ xserver-xlibre-xserver-25.1.0/hw/xfree86/loader/loader.c	2025-12-25 23:53:55.116292868 +0200
@@ -52,6 +52,7 @@
 
 #include <string.h>
 #include "os.h"
+#include "list.h"
 #include "loader.h"
 #include "loaderProcs.h"
 
@@ -72,6 +73,15 @@
 Bool LoaderIgnoreABI;
 Bool is_nvidia_proprietary;
 
+/* ABI for a module with this canonical name is ignored */
+typedef struct {
+    struct xorg_list entry;
+    char * name;
+} LoaderIgnoreABIItem;
+
+/* List of such modules */
+struct xorg_list LoaderIgnoreABIList;
+
 void
 LoaderInit(void)
 {
@@ -96,12 +106,19 @@
     LoaderIgnoreABI = FALSE;
     is_nvidia_proprietary = FALSE;
     LoaderInitPath();
+    xorg_list_init(&LoaderIgnoreABIList);
 }
 
 void
 LoaderClose(void)
 {
     LoaderClosePath();
+    LoaderIgnoreABIItem *item, *next;
+    xorg_list_for_each_entry_safe(item, next, &LoaderIgnoreABIList, entry) {
+        xorg_list_del(&item->entry);
+        free(item->name);
+        free(item);
+    }
 }
 
 /* Public Interface to the loader. */
@@ -161,25 +178,70 @@
         dlclose(handle);
 }
 
+/*
+ * The functions below are necessary to load some modules, e.g., nvidia proprietary drivers,
+ * regardless of their ABI versions
+ */
+
+
+/* ABI versions for all modules will be ignored */
 void
 LoaderSetIgnoreAllABI(void)
 {
     LoaderIgnoreAllABI = TRUE;
 }
 
+/* Check whether ABI for a currently loaded module is ignored and set the flag respectively */
 Bool
 LoaderGetAndFlagIgnoreABI(const char *name)
 {
-    if (LoaderIgnoreAllABI)
+    LoaderIgnoreABIItem *item;
+
+    if (LoaderIgnoreAllABI) {
         LoaderIgnoreABI = TRUE;
-    return LoaderIgnoreABI;
+        return TRUE;
+    }
+
+    xorg_list_for_each_entry(item, &LoaderIgnoreABIList, entry) {
+        if (!strcmp(item->name, name)) {
+            LoaderIgnoreABI = TRUE;
+            return TRUE;
+        }
+    }
+
+    LoaderIgnoreABI = FALSE;
+    return FALSE;
 }
 
+/* Add the module with this name to the list of modules with ignored ABI */
 void
 LoaderSetIgnoreABI(const char *name)
 {
+    LoaderIgnoreABIItem *item;
+
+    if (LoaderIgnoreAllABI)
+        return;
+
+    xorg_list_for_each_entry(item, &LoaderIgnoreABIList, entry) {
+        if (!strcmp(item->name, name))
+            goto out;
+    }
+
+    item = malloc(sizeof(LoaderIgnoreABIItem));
+    if (item)
+        item->name = strdup(name);
+    if (item && item->name)
+        xorg_list_add(&item->entry, &LoaderIgnoreABIList);
+    else {
+        LogMessage(X_ERROR, "Failed to allocate memory to store ignore ABI for module %s\n", name);
+        if (item)
+            free(item);
+    }
+
+  out:
 }
 
+/* These two functions are called by legacy nvidia drivers */
 Bool
 LoaderShouldIgnoreABI(void)
 {
--- xserver-xlibre-xserver-25.1.0/hw/xfree86/common/xf86platformBus.c.orig	2025-12-21 16:53:30.000000000 +0200
+++ xserver-xlibre-xserver-25.1.0/hw/xfree86/common/xf86platformBus.c	2025-12-26 00:03:54.299312042 +0200
@@ -267,6 +267,7 @@
     }
 
     for (i = 0; i < xf86_num_platform_devices; i++) {
+        struct xf86_platform_device *dev = &xf86_platform_devices[i];
         char *busid = xf86_platform_odev_attributes(i)->busid;
 
         if (pci && busid && (strncmp(busid, "pci:", 4) == 0)) {
@@ -350,6 +351,39 @@
                 }
                 /* Otherwise global module search path is left unchanged */
             }
+
+            if (xf86CheckBoolOption(cl->option_lst, "IgnoreABI", FALSE)) {
+                if (cl->driver) {
+                    LogMessageVerb(X_CONFIG, 1, "OutputClass \"%s\" sets driver \"%s\" to ignore ABI for \"%s\" managed by %s\n",
+                               cl->identifier, cl->driver, dev->attribs->path, dev->attribs->driver);
+                    if (*(cl->driver)) LoaderSetIgnoreABI(cl->driver);
+                    if (cl->modules) {
+                        LogMessageVerb(X_CONFIG, 1, "    and for modules \"%s\" as well\n",
+                                cl->modules);
+                        XNFasprintf(&copy, "%s", cl->modules);
+                        curr = copy;
+                        while ((curr = strtok_r(curr, ",", &next))) {
+                            if (*curr) LoaderSetIgnoreABI(curr);
+                            curr = NULL;
+                        }
+                        free(copy);
+                    }
+                } else if (cl->modules) {
+                    LogMessageVerb(X_CONFIG, 1, "OutputClass \"%s\" sets modules \"%s\" to ignore ABI for \"%s\" managed by %s\n",
+                               cl->identifier, cl->modules, dev->attribs->path, dev->attribs->driver);
+                    XNFasprintf(&copy, "%s", cl->modules);
+                    curr = copy;
+                    while ((curr = strtok_r(curr, ",", &next))) {
+                        if (*curr) LoaderSetIgnoreABI(curr);
+                        curr = NULL;
+                    }
+                    free(copy);
+                } else {
+                    LogMessageVerb(X_CONFIG, 1, "OutputClass \"%s\" requires to ignore ABI for \"%s\" managed by %s\n"
+                            "   but neither \"Driver\" nor \"Module\" is set\n",
+                            cl->identifier, dev->attribs->path, dev->attribs->driver);
+                }
+            }
         }
     }
 
